### PR TITLE
[wasm] Make runtime_is_initialized promise callbacks one-shot

### PR DESCRIPTION
**Pull request by lambdageek**
Throw if `runtime_is_initialized_resolve` or `runtime_is_initialized_reject` is called more than once

### CODE REVIEW

**Comment by kg** [comment]
You don't need to do it, but it's worth considering whether we should also explicitly throw if resolve() is called after reject(), or vice-versa. Looking at the spec it appears that the normal behavior for mistakes like this (and for double-reject) is silently ignoring the call, and we want to throw.

image

Also, according to documentation if the executor (the function being passed resolve/reject) throws it will reject the promise, so that would make calling reject() after that an error. Thankfully in this case it's not possible for our executor to throw, but it's worth knowing if you use this pattern in the future :)

**Review by kg — APPROVED** [review]

**Review by pavelsavara — APPROVED** [review]

**Comment by lambdageek** [comment]
The original motivation for this PR is that I was confused by the error handling in `startup.ts` and was pretty convinced that we would call `reject` multiple times.  Multiple-`accept` isn't really even a suspicion on any control flow path AFAICT.

I guess it might be worthwhile to also guard against `resolve`-after-`reject` (and vice versa, though it seems unlikely).

> You don't need to do it, but it's worth considering whether we should also explicitly throw if resolve() is called after reject(), or vice-versa. 
> …
> Also, according to documentation if the executor (the function being passed resolve/reject) throws it will reject the promise, so that would make calling reject() after that an error. Thankfully in this case it's not possible for our executor to throw, but it's worth knowing if you use this pattern in the future :)

I think there's probably value in making a `GuardedPromise<T>` (or something), but I'm not sure if dotnet is the place for it.

**Comment by pavelsavara** [comment]
What would be the problem with calling reject multiple times, even if that happened ? It would reject the promise only once, right ? Do we need to learn about the second time ?

File: src/mono/wasm/runtime/guarded-promise.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/// A Promise<T> that guards against multiple-resolve, multiple-reject, reject-after-accept and accept-after-reject.
```
Comment (lambdageek) [inline src/mono/wasm/runtime/guarded-promise.ts:1]:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

/// A Promise<T> that guards against multiple-resolve, multiple-reject, reject-after-accept and accept-after-reject.
```

**Comment by pavelsavara** [comment]
> After line 345 we should really return instead of falling through to `bindings_lazy_init`

AFAIK `wasm_exit` would throw or really exit in nodejs.

**Comment by pavelsavara** [comment]
> > > After line 345 we should really return instead of falling through to `bindings_lazy_init`
> > 
> > 
> > AFAIK `wasm_exit` would throw or really exit in nodejs.
> 
> I'm talking about the browser case - we reject the promise, skip the `if (ENVIRONMENT_IS_SHELL)` and fall through to `bindings_lazy_init`.

I see, I guess we should throw explicitly there.

**Pull request merged**